src/compiler.ts

```diff
@@ -7220,18 +7220,21 @@ export class Compiler extends DiagnosticEmitter {
     var stub = original.varargsStub;
     if (stub) return stub;
 
+    // this stub relies on the existence of the argumentsLength global
+    this.ensureArgumentsLength();
+
     var originalSignature = original.signature;
     var originalParameterTypes = originalSignature.parameterTypes;
     var originalParameterDeclarations = original.prototype.functionTypeNode.parameters;
     var returnType = originalSignature.returnType;
-    var isInstance = original.is(CommonFlags.INSTANCE);
+    var thisType = originalSignature.thisType;
```

Comment:
Rather than asking if something is an instance to infer if it has a this argument, just check directly if it has a this type.

Now it recognizes when a function is a closure and must be called with a context